TypeScript_ESLint.snb.md

```diff
@@ -0,0 +1,131 @@
+
+Let's dive into the definition of this rule in source. It is short enough to include here in its entirety:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts
+
+The rule definition contains metadata at the top and the actual logic of the rule below. Let's zoom in on that logic:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L22-41
+
+There may be some unfamiliar syntax for some here. This code might look like it's defining a function called `ForInStatement` within an object declaration, but it's really just a terse variant of declaring a function value for an object key, like this:
+
+```typescript
+{
+    ForInStatement: (node): void => { ... }
+}
+```
+
+This key will be used to match this rule with the name of an AST node type in the ESLint AST, so that the node argument will always have the type that the rule targets:
+
+https://sourcegraph.com/npm/typescript-eslint/types@540cca215af38ab1a49b6cde8bcb492f4b6e53c5/-/blob/dist/generated/ast-spec.d.ts?L75
+
+## Type-checking
+
+One distinction between TypeScript ESLint and vanilla ESLint is that we have access to the TypeScript type checker and can therefore define rules that make use of type information.
+
+The `for-in-array` rule uses the type checker to determine if we are dealing with an array type:
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L23-24
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L32-35
+
+## Dual ASTs
+
+In addition to invoking the type checker, the `for-in-array` rule also makes use of **two ASTs**:
+
+1. the ESTree representation used by ESLint
+2. the TypeScript AST emitted by the TypeScript compiler.
+
+The `node` value passed as an argument to the rule function is from AST #1, an instance of [`TSESTree.Node`](https://sourcegraph.com/npm/typescript-eslint/types/-/blob/dist/generated/ast-spec.d.ts?L641):
+
+The `originalNode` value is from AST #2, an instance of ([`TSNode`](https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@3eab889022c9d1617f275017d6951f663ea57f24/-/blob/packages/typescript-estree/src/ts-estree/ts-nodes.ts?L18:13#tab=references)):
+
+https://sourcegraph.com/github.com/typescript-eslint/typescript-eslint@HEAD/-/blob/packages/eslint-plugin/src/rules/no-for-in-array.ts?L25#tab=def
+
+Why two ASTs instead of one? Because each provides a distinct set of useful functionality.
+
+1. The `TSESTree` representation is used by ESLint internally and enables the plugin to reuse much of the code and linting framework provided by ESLint.
+2. The `TSNode` representation is emitted by the TypeScript compiler and contains type information.
```

Comment:
I want to say the `TSNode`s don't have type information themselves? I might be wrong on how the internal representation is set up. But, the external API makes it seem that way, so that's how I'd suggest phrasing it.

```suggestion
2. The `TSNode` representation is emitted by the TypeScript compiler and is what the type checker's APIs expect as inputs
```

Reply:
you are correct, ts-ast, has no type data, we have to call checker service to get them, 

(to be clear after first call to checker private cache like structure is added to ts-ast)